src/v/cluster/shard_table.h

```diff
@@ -104,6 +108,21 @@ class shard_table final {
           log_rev);
         _ntp_idx.erase(ntp);
         _group_idx.erase(g);
+
+        _notification_list.notify(ntp, g, std::nullopt);
+    }
+
+    using change_cb_t = ss::noncopyable_function<void(
+      const model::ntp& ntp,
+      raft::group_id g,
+      std::optional<ss::shard_id> shard)>;
+
+    notification_id_type register_notification(change_cb_t&& cb) {
```

Comment:
no need to take r-value reference here

Reply:
I can name a couple of reasons to use r-value reference here:
1. Hinting move semantics to the caller without making them look into the type definition. Might seem useless here, as how do they know the callback signature, but what if they already have the callback value and are making an attempt to reuse it they will probably realise it won't work a bit earlier.
2. Optimizing out actual moves https://godbolt.org/z/qbcManb57. They are probably cheap but not zero-cost.

Any reasons to pass by value?

Reply:
the primary reason is that (1) it's already non-copyable so you have to move it--no benefit in semantics and (2) it's not in a hot path, right? so copying 40 bytes during registration is nothing. anyway, nothing wrong with using && but it's nice to default to value semantics where possible. no need to change or anything.